core/lib/vm/src/old_vm/oracles/storage.rs

```diff
@@ -321,18 +325,24 @@ impl<S: WriteStorage, H: HistoryMode> VmStorageOracle for StorageOracle<S, H> {
 
 /// Returns the number of bytes needed to publish a slot.
 // Since we need to publish the state diffs onchain, for each of the updated storage slot
-// we basically need to publish the following pair: (<storage_key, new_value>).
-// While new_value is always 32 bytes long, for key we use the following optimization:
+// we basically need to publish the following pair: (<storage_key, compressed_new_value>).
+// For key we use the following optimization:
 //   - The first time we publish it, we use 32 bytes.
 //         Then, we remember a 8-byte id for this slot and assign it to it. We call this initial write.
-//   - The second time we publish it, we will use this 8-byte instead of the 32 bytes of the entire key.
-//         So the total size of the publish pubdata is 40 bytes. We call this kind of write the repeated one
-fn get_pubdata_price_bytes(_query: &LogQuery, is_initial: bool) -> u32 {
+//   - The second time we publish it, we will use the 4/5 byte representation of this 8-byte instead of the 32
+//     bytes of the entire key.
+// For value compression, we use a metadata byte which holds the length of the value and the operation from the
+// previous state to the new state, and the compressed value. The maxiumum for this is 33 bytes.
+// Total bytes for initial writes then becomes 65 bytes and repeated writes becomes 38 bytes.
+fn get_pubdata_price_bytes(query: &LogQuery, is_initial: bool) -> u32 {
     // TODO (SMA-1702): take into account the content of the log query, i.e. values that contain mostly zeroes
     // should cost less.
+    let compressed_value_size =
+        compress_with_best_strategy(query.read_value, query.written_value).len() as u32;
```

Comment:
you should compare not `.read_value` and `.written_value`, but between the value that was at the very start of the batch in this slot and the one that is being written to this slot

for instance, if we incremented by 255 (which is add with 1 byte) and then we incremented once more by 255 (the diff between read_value and written_value will be add with 1 byte also). The `value_update_price` will see that the user has already paid for 1 byte, while actually the user needs to pay for the additional one byte created (since the final diff requires 2 bytes). 

If it is not too hard, I would also suggest creating a test for such case

Reply:
instead of that, could we check to see if the `written_value` > `read_value` and charge for the difference? if it's `<=` then we set the cost be `0`